src/frpcserverproxy.cc

```diff
@@ -241,19 +259,221 @@ Marshaller_t* ServerProxyImpl_t::createMarshaller(HTTPClient_t &client) {
     return marshaller;
 }
 
+/**
+ * Multifaceted data structure used to cache ServerProxy_t objects.
+ * Maintains map for lookup by url, and timer queues for removing old objects.
+ * The map and queues are interlinked. Timer queues are organized in powers of
+ * two.
+ */
+class ProxyCache_t {
+    struct LessURL_t {
+        bool operator()(const URL_t &lhs, const URL_t &rhs) const {
+            if (lhs.port == rhs.port && lhs.path == rhs.path) {
+                return lhs.host < rhs.host;
+            }
+
+            if (lhs.port == rhs.port) {
+                return lhs.path < rhs.path;
+            }
+
+            return lhs.port < rhs.port;
+        }
+    };
+
+    class TimeoutListHead;
+    using Time = uint64_t;
+    using Key_t = URL_t;
+    using Value_t = std::unique_ptr<TimeoutListHead>;
+    using Map_t = std::multimap<Key_t, Value_t, LessURL_t>;
+    static constexpr size_t TIMEOUT_QUEUES = 20;
+
+    ProxyCache_t(int timeout) : timeout(timeout) {}
+
+    static Time gettimemilliseconds() {
+        timeval tv;
+        gettimeofday(&tv, nullptr);
+        return tv.tv_sec * 1000 + ((tv.tv_usec + 500) / 1000);
+    };
+
+    /** Intrusive linked-list structure for timer queues. */
+    class TimeoutListHead {
+        friend class ProxyCache_t;
+        TimeoutListHead() : next(this), prev(this) {}
+
+    public:
+        TimeoutListHead(std::unique_ptr<ServerProxyImpl_t> impl)
+            : impl(std::move(impl)), next(this), prev(this)
+        {}
+        ~TimeoutListHead() { delink(); }
+
+        void delink() {
+            next->prev = prev;
+            prev->next = next;
+            prev = next = this;
+        }
+
+        bool is_linked() const { return next != this; }
+
+        std::unique_ptr<ServerProxyImpl_t> impl;
+
+    protected:
+        TimeoutListHead *next;
+        TimeoutListHead *prev;
+        Map_t::iterator self;
+        Time next_setoff_time = 0;
+        Time final_setoff_time = 0;
+    };
+
+    /** Inserts timer into correct queue */
+    void schedule(TimeoutListHead &t, Time now) {
+        if (t.is_linked()) {
+            t.delink();
+        }
+
+        // find the correct timer queue
+        auto timeout = t.final_setoff_time - now;
+        t.next_setoff_time = t.final_setoff_time;
+        size_t i = 0;
+        for (; i < TIMEOUT_QUEUES - 1; ++i) {
+            timeout >>= 1;
+            if (timeout == 0) {
+                timeout = 1 << i;
+                t.next_setoff_time = now + timeout;
+                break;
+            }
+        }
+
+        // insert into the queue
+        auto &queue = timeouts[i];
+        queue.prev->next = &t;
```

Comment:
Originally there was one. It was used only at this place and was rather confusing what gets inserted after what. If it was named `link_before` or `insert_before` it may be okay but the asymmetry of roles between header nodes and real timeout nodes does not play nicely with that.